### ISSUE TITLE
Multiple services for the same workload in Mixer (telemetry and policy)

### ISSUE THREAD

**Issue by dicarlo2** [istio/istio#11578]
**Describe the bug**
In order to workaround issues with referring and connecting to statefulset pods directly (see https://github.com/istio/istio/issues/10659), we create a service per pod in the statefulset. For example:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: thanos-store
  namespace: monitoring-potato-beetle
  labels:
    app: thanos-store
spec:
  selector:
    app: thanos-store
  ports:
  - name: http-prom
    port: 10902
  - name: grpc
    port: 10901
---
apiVersion: v1
kind: Service
metadata:
  name: thanos-store-0
  namespace: monitoring-potato-beetle
  labels:
    app: thanos-store
spec:
  selector:
    app: thanos-store
    statefulset.kubernetes.io/pod-name: thanos-store-0
  ports:
  - name: http-prom
    port: 10902
  - name: grpc
    port: 10901
---
apiVersion: v1
kind: Service
metadata:
  name: thanos-store-1
  namespace: monitoring-potato-beetle
  labels:
    app: thanos-store
spec:
  selector:
    app: thanos-store
    statefulset.kubernetes.io/pod-name: thanos-store-1
  ports:
  - name: http-prom
    port: 10902
  - name: grpc
    port: 10901
```

One minor issue this causes is that when using authz every service name must be included in the `ServiceRole`:

```yaml
apiVersion: rbac.istio.io/v1alpha1
kind: ServiceRole
metadata:
  name: thanos-query.monitoring-potato-beetle-thanos-store.monitoring-potato-beetle
  namespace: monitoring-potato-beetle
  labels:
    app: thanos-query
spec:
  rules:
  - services:
    - thanos-store.monitoring-potato-beetle.svc.cluster.local
    - thanos-store-0.monitoring-potato-beetle.svc.cluster.local
    - thanos-store-1.monitoring-potato-beetle.svc.cluster.local 
    constraints:
    - key: destination.port
      values:
      - "10901"
    methods:
    - POST
```

The much worse issue, however, is that the service name used in telemetry appears to change over time:

<img width="1078" alt="screenshot 2019-02-06 16 04 47" src="https://user-images.githubusercontent.com/1311014/52382172-0f495280-2a29-11e9-84ab-f04ada4936aa.png">

This screenshot shows a similarly setup service `grafana` with one additional service for the single workload (pod) `grafana-0`:

<img width="1073" alt="screenshot 2019-02-06 15 57 54" src="https://user-images.githubusercontent.com/1311014/52382128-f3de4780-2a28-11e9-8292-6eab3c444fd3.png">

**Expected behavior**

One service name is used so that telemetry is consistent. Perhaps there could be an annotation that marks a service as the canonical service to use in policy and telemetry.

**Steps to reproduce the bug**
See description of setup.

**Version**
release-1.1-20190130-09-16

**Installation**
Mostly equivalent to helm with flat multicluster and cni.

**Environment**
GKE

**Related**
https://github.com/istio/istio/issues/1459
https://github.com/istio/istio/issues/10659

**Comment by douglas-reid** [comment]
The `destination_service` label is taken from the `destination.service.host` attribute, so I believe this behavior is expected.

Would using `destination_app` be more appropriate in your queries?  That is already using a "special" label (`destination.labels["app"]`).

**Comment by dicarlo2** [comment]
I had some time to look into this a bit more today - you're correct that it comes from the `destination.service.host` attribute in the envoy filter's route config. After a bit of digging through the code, I realized that this is mostly expected, when istio encounters duplicate services for a particular workload/port it simply picks one. However, I'd argue that it should pick one consistently (e.g. by sorting service names, or some other heuristic). For example, here are two dumps of the relevant envoy listener for the above grafana case, one where [`grafana-0`](https://gist.github.com/dicarlo2/edb2aa595e83e29d0d4e852360fb7a65) is selected as the destination service and one where [`grafana`](https://gist.github.com/dicarlo2/c0359855079e951b29aa370ed005faab) is selected. Here's the particular bit of code: 
https://github.com/istio/istio/blob/92f3d9716f1bf6f35befc07fd5deb1974b381f16/pilot/pkg/networking/core/v1alpha3/listener.go#L489-L509

If I were to take a guess, it's likely because somewhere along the line there's a map iteration which randomizes the order, but I wasn't able to pinpoint exactly where this was happening.

Another thing I noticed is that (at least on `release-1.1-20190130-09-16`) `ProxyStatus` is never logged, so configuration issues are relatively hard to detect by simply looking at pilot logs. Seems like it's probably related to the move to incremental updates, as the logging occurs only for the full push codepath.

In summary, I'd suggest two things:

1. Choose the conflicting listener in a consistent manner
2. Add periodic logging to pilot for `ProxyStatus` to make discovering configuration issues a bit easier.

I resolved the issue by ensuring there are no conflicts, which happens to work in my case. The ports which require pod-specific requests are always made with pod-specific requests so they don't need to be in the main service definition (and visa versa). With that said, I can envision scenarios where this is not the case, so I'd suggest fixing 1/. I'll leave this issue open to track fixing 1/. If you disagree, feel free to close.
